Language Ref Python Builtin Functions and Values (#52830)
pytorch/pytorch: docs/source/jit_python_values.rst

.. contents::
    :local:
    :depth: 2

.. testsetup::

    # These are hidden from the docs, but these are necessary for `doctest`
    # since the `inspect` module doesn't play nicely with the execution
    # environment for `doctest`
    import torch

    original_script = torch.jit.script
    def script_wrapper(obj, *args, **kwargs):
        obj.__module__ = 'FakeMod'
        return original_script(obj, *args, **kwargs)
    torch.jit.script = script_wrapper

    original_trace = torch.jit.trace
    def trace_wrapper(obj, *args, **kwargs):
        obj.__module__ = 'FakeMod'
        return original_trace(obj, *args, **kwargs)
    torch.jit.trace = trace_wrapper

.. _python-builtin-functions-values-resolution:

Resolution Rules
================

> When given a Python value, TorchScript attempts to resolve them in following five different ways:

Reviewer: ```suggestion
When given a Python value, TorchScript attempts to resolve it in following five different ways:
```

Author: Done.

* Compilable Python Implementation: 

>     * When a Python value is backed by a Python implementation that can be compiled by TorchScript, TorchScript compiles and uses the underlying Python function.

Reviewer: What if the implementation is not a function? I'm confused by the words `underlying Python function`.

Author: Changed to "underlying Python implementation"

    * Example: ``torch.jit.Attribute``
* Op Python Wrapper: 
    * When a Python value is a wrapper of a native PyTorch op, TorchScript emits corresponding operator.
    * Example: ``torch.jit._logging.add_stat_value``
* Python Object Identity Match: 
    * For a limited set of ``torch.*`` API calls (in the form of Python values) that TorchScript supports, TorchScript attempts to match a Python value against each item in the set. 

>     * When matched, TorchScript generates a corresponding SugaredValue instance that contains lowering logic for these values.

Reviewer: There's no explanation in other parts of the spec what `SugaredValue` is, so this might be confusing.
```suggestion
    * When matched, TorchScript generates a corresponding ``SugaredValue`` instance that contains lowering logic for these values.
```

Author: Added code formatting per suggested fix.

Are there any existing doc on what SugaredValue is? if not, I could add a link pointing to it, otherwise, I could add a few sentences here describing very roughly on what a SugaredValue is.

    * Example: ``torch.jit.isinstance``
* Name Match: 
    * For Python built-in functions and constants, TorchScript identifies them by name, and creates a corresponding SugaredValue instance that implements their functionality.
    * Example: ``all()``
* Value Snapshot: 
    * For Python values from unrecognized modules, TorchScript attempts to take a snapshot of the value and converts to a constant in the graph of the function(s) or method(s) being compiled
    * Example: ``math.pi``

.. _python-builtin-functions-support:

> Python Builtin Functions Support

Reviewer: As I mentioned before, this would be better as a table.

Author: Done.

``memoryview()``
^^^^^^^^^^^^^^^^
* Support Level: 
    * None

``min()``
^^^^^^^^^
* Support Level: 
    * Partial
* Notes:
    * ``key`` argument not supported

``next()``
^^^^^^^^^^
* Support Level: 
    * None

``object()``
^^^^^^^^^^^^
* Support Level: 
    * None

``oct()``
^^^^^^^^^
* Support Level: 
    * Partial
* Notes: 
    * Only supports ``Int``-type input

``open()``
^^^^^^^^^^
* Support Level: 
    * None

``ord()``
^^^^^^^^^
* Support Level: 
    * Partial
* Notes: 
    * Only ASCII character set is supported
    
``pow()``
^^^^^^^^^^
* Support Level: 
    * Full

``print()``
^^^^^^^^^^^
* Support Level: 
    * Full
* Notes:
    * ``separate``, ``end`` and ``file`` arguments are not supported

``property()``
^^^^^^^^^^^^^^
* Support Level: 
    * None

``range()``
^^^^^^^^^^^
* Support Level: 
    * Full

``repr()``
^^^^^^^^^^
* Support Level:
    * None

``reversed()``
^^^^^^^^^^^^^^
* Support Level:
    * None

``round()``
^^^^^^^^^^^^^^
* Support Level:
    * Partial
* Notes:
    * ``ndigits`` argument is not supported

``set()``
^^^^^^^^^^
* Support Level:
    * None

``setattr()``
^^^^^^^^^^^^^
* Support Level:
    * Partial
* Notes:

>     * Only supports assignment to existing attribute, ie. doesn't support creating new attribute on object

Reviewer: ```suggestion
    * Only supports assignment to existing attribute, i.e. it doesn't support creating new attribute on object
```
